src/dune/build_system.ml

```diff
@@ -1340,11 +1351,49 @@ end = struct
       | exception Unix.Unix_error (ENOENT, _, _) -> ()
       | () -> () )
 
+  let execute_action_until_all_deps_ready ~context ~env ~targets ~rule_loc
+    action =
+    let open Dune_action.Protocol in
+    let provided_dependencies_set = ref Dependency.Set.empty
+    and stages = ref Appendable_list.empty in
+    let rec loop () =
+      let* result =
+        Action_exec.exec ~context ~env ~targets ~rule_loc
+          ~provided_dependencies:!provided_dependencies_set action
+      in
+      match result with
+      | Done -> Fiber.return (Appendable_list.to_list !stages)
+      | Need_more_deps deps ->
+        let deps_to_build = Dependency.Map.values deps |> Dep.Set.of_list in
+        (stages := Appendable_list.(!stages @ singleton deps_to_build));
+        let* () = build_deps deps_to_build in
+        provided_dependencies_set :=
+          Dependency.Set.union !provided_dependencies_set
+            (Dependency.Set.of_keys deps);
+        loop ()
+    in
+    loop ()
+
+  let compute_rule_digest (rule : Internal_rule.t) ~deps ~action ~sandbox_mode
+    =
+    let targets_as_list = Path.Build.Set.to_list rule.targets in
+    let env = Internal_rule.effective_env rule in
+    let trace =
+      ( Dep.Set.trace deps ~sandbox_mode ~env ~eval_pred
+      , List.map targets_as_list ~f:(fun p -> Path.to_string (Path.build p))
+      , Option.map rule.context ~f:(fun c -> c.name)
+      , Action.for_shell action )
+    in
+    Digest.generic trace
+
+  let compute_dependencies_digest deps ~sandbox_mode ~env ~eval_pred =
```

Comment:
By the way I think `Dep.Set.trace` is doing something suspicious regarding `Glob` deps. We might want to look at it more carefully and convince ourselves that it's good (or bad).

Reply:
My understanding is that:
* `Dep.Set.trace` will just call `Dep.trace` on each element of the set 
(and if only `Dep.t`'s compare is reasonable then we should always get the same order)
* `Dep.trace` on `File_selector` variant will just call `eval_pred` getting the set of targets and calling `trace_file` on each target
* `trace_file` just computes the file digest (and return pair of it and the filename)
* `eval_pred` computes the set of targets by taking the `dir` associated with file selector, getting targets from that dir by calling `targets_of` and then just filtering it by `File_selector.test` function

Reply:
I've always found it suspicious that it computes the digest of a set of files instead of treating glob as an observable unit in itself.
Consider an action `ls -1 foo`.
Let's say we have two runs:

1. deps are empty; directory foo contains file "x"

now dune will remember that when no dependencies are provided, "ls" returns "x"

2. deps are correctly specified ([Glob "foo/*"]), directory foo is empty

Dune will expand the glob, find that it's empty so we have the same empty deps as the first time, so it's OK to re-use the same result...

In the end, even though the deps are correctly specified at step (2), dune is still confused.

Maybe dynamic-run does not make the problem worse though. I'm not sure.

Reply:
Good catch! This problem is independent from dune-dynamic, created #2604  for fixing this.